Implements proxying feature that runs haproxy

longshorej opened the pull request:
Implements proxying feature that runs haproxy. This is part of the default set of features that can be disabled with `--no-default-features`. Features now have a `stop` method that will be called when `sandbox` stop is run, and proxying takes advantage of that.

longshorej commented on conductr_cli/sandbox_run.py:
```diff
@@ -42,15 +42,14 @@ def run(args):
             result = feature.start()
             feature_results += result
 
+            if isinstance(feature, sandbox_features.ProxyingFeature):
```
This I'm a bit iffy about, because Proxying is a "special" feature.

huntc replied:
Yeah, if feels wrong to do a special test. I thought that features had some facility to contribute to the output. Perhaps look at the monitoring feature for inspiration.

fsat replied:
I'd suggest introduce a property `self.print_start_output = True` for the `ProxyingFeature`, and `self.print_start_output = False` for everything else. This way you can just check if `feature.print_start_output` is `True`, and print the output accordingly. Wdyt?

longshorej replied:
I think your suggestion to add a property to the feature is right but I'm thinking we can generalize it a bit.

As an idea, maybe features also have a `provides` property (e.g. the `proxy` and `log` class) such that only a single feature can be enabled at a time that has that `provides` entry. Then, this check simply becomes, were any features with `provides` `proxy` started? If so, we alter the service URL displayed. In the future, this would allow multiple feature implementations of the proxying concept and `sandbox_run_jvm.py` wouldn't care which was actually enabled.

Similarly, this would generalize the eslite/es logic. We could throw an error if two features were specified that both provided logging.

Just for reference, this boolean (`print_proxy_output`) is used as the following:

```
               for feature_result in feature_results:
                    if is_proxy_started:
                        uri = '{}:{}'.format(run_result.host, feature_result.port)
                    else:
                        uri = '{}:{}/services/{}'.format(run_result.host, DEFAULT_SERVICE_LOCATOR_PORT,
```

longshorej replied:
I went ahead with my idea here. Features now have a `provides` mechanism to provide any arbitrary string. No two enabled features can provide that same value -- this allows us to ensure only one logging feature is enabled, or in the future only one proxying feature. Let me know what you guys think of this!

fsat commented on conductr_cli/constants.py:
```diff
@@ -2,6 +2,8 @@
 
 CONDUCTR_SCHEME = 'CONDUCTR_SCHEME'
 
+CONDUCTR_BUNDLE_PROXY_PORT = 9000  # The default proxy port for bundles.
```
I think we would be able to make do without this constant. Refer to the comments below.

longshorej replied:
Done.

fsat commented on conductr_cli/sandbox_features.py:
```diff
@@ -27,6 +28,73 @@ def __init__(self, name, port):
+            if started:
+                log.info('HAProxy has been started')
+                log.info('Your Bundles are by default accessible on:')
+                log.info('  {}:{}'.format(self.host, CONDUCTR_BUNDLE_PROXY_PORT))
```
I think `CONDUCTR_BUNDLE_PROXY_PORT` should be replaced in lieu of `self.bundle_http_port` as it's possible for the user to set bundle port apart from default `9000` and this value is kept in `self.bundle_http_port`.

longshorej replied:
Done

fsat commented on conductr_cli/test/test_sandbox_features.py:
```diff
@@ -1,7 +1,7 @@
 from unittest.mock import call, patch, MagicMock
 from conductr_cli.sandbox_features import VisualizationFeature, LiteLoggingFeature,\
-    LoggingFeature, MonitoringFeature, \
+    LoggingFeature, MonitoringFeature, ProxyingFeature, \
```
I'd suggest more coverage for `ProxyingFeature` in these areas:

- Output printout of during startup including when docker is enabled vs docker not enabled.
- Asserting the input argument and the call to `sandbox_proxy` during feature start depending of the input argument passed into `ProxyingFeature`'s `configure` method.
- Asserting the input argument and the call to `sandbox_proxy` during feature stop.

longshorej replied:
Good point - coverage was added for all of these cases except the first point. I moved the call to `sandbox_proxy.start` and there are tests on that method already to check the docker logic, so I think we're okay from a coverage standpoint.

fsat commented on conductr_cli/test/test_sandbox_run.py:
```diff
@@ -40,6 +40,26 @@ def default_general_args(self, container_name):
+
+        def start(self):
+            return []
+
```
I can't seem to find the usage of `DummyFeature1` and `DummyFeature2`.

If they are meant to be used to assert that `configure` and `start` has been called, perhaps can be replaced with `MagicMock`, i.e.:

```
def create_dummy_features():
  feature = MagicMock()
  feature.configure = MagicMock()
  feature.start = MagicMock()
  return feature

feature_1 = self.create_dummy_features()
feature_2 = self.create_dummy_features()
features = [feature_1, feature_2] # use this as the return_value of mock_collect_features

feature_1.configure.assert_called_once_with(...)
feature_1.start.assert_called_once_with(...)
```

If they are indeed not used, can we remove them please?

longshorej replied:
Done -- sorry, leftover from earlier iterations that I forgot to remove.

fsat reviewed (changes requested):
Nice work!

I have some feedback about Proxying being a special feature, as well as some question about the new constant and test coverage for the new proxying feature.

markusjura commented on conductr_cli/sandbox_features.py:
```diff
@@ -27,6 +28,73 @@ def __init__(self, name, port):
+        self.bundle_http_port = None
+        self.proxy_ports = None
+
+    def configure(self, args, run_result):
```
The name `configure` is quite general and does not indicate when it is called. What about `postConductrStart`?

longshorej replied:
Done - per python conventions and other methods beginning with `conductr_`, went with `conductr_post_start`

markusjura reviewed (dismissed):
Looking good. Just one additional comment in regards to method naming.

longshorej commented:
Okay, thanks @markusjura and @fsat for the feedback. I've incorporated it in the latest commit. To deal with the special case of the proxy being started affecting the URI calculation in `sandbox_run_jvm.log_run_attempt`, I introduced another concept for a `Feature` called `provides`. This is a list of arbitrary strings. `sandbox_run_jvm.log_run_attempt` then checks if any features provide `proxying` to calculate which URL to show.

This would allow us to swap out several different proxy features in the future, and allows us to introduce mutual exclusion for `logging` vs `lite-logging`.

I think it turned out nicely but let me know what you guys think.

fsat commented:
@longshorej - Thanks for incorporating my feedback. I've reviewed the PR, and I like the changes so far!

I do have a question about the `ProxyingFeature` test which I think it'll be faster if we can talk quickly after standup tomorrow.

longshorej committed:
Implements proxying feature that runs haproxy
